### PR TITLE
TextField - replace withFieldState HOC with useFieldState hook

### PR DESCRIPTION
**Pull request by ethanshar**
I realize we can accomplish the functionality of `withFieldState` HOC with a hook.

Since it doesn't affect directly the render output and it just provides logic and data and because the whole TextField is implemented using function components, there's no reason not to use hooks. 

Also, in this PR I declare the "validate" typings in the main component as well (even though it's repetitive) just for react-docgen

### CODE REVIEW

File: src/incubator/TextField/useFieldState.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      _.invoke(props, 'onChangeText', text);
+
+      if (validateOnChange) {
+        validateField(text);
+      }
+    },
+    [props.onChangeText, validateOnChange]
+  );
+
+  const fieldState = useMemo(() => {
```
Comment (M-i-k-e-l) [inline src/incubator/TextField/useFieldState.ts:80]: What is this used for? Form?

File: src/incubator/TextField/useFieldState.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+export default function useFieldState({
+  validate,
+  validateOnBlur,
+  validateOnChange,
+  validateOnStart,
+  ...props
+}: FieldStateProps) {
+  const [value, setValue] = useState(props.value);
+  const [isFocused, setIsFocused] = useState(false);
```
Comment (M-i-k-e-l) [inline src/incubator/TextField/useFieldState.ts:26]: Do we have to have `isFocused`? Each time a `TextField` is focused\blurred it is re-rendered.

File: src/incubator/TextField/useFieldState.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
+import {useCallback, useState, useEffect, useMemo} from 'react';
+import {TextInputProps} from 'react-native';
+import _ from 'lodash';
+import validators from './validators';
+
+export type Validator = Function | keyof typeof validators;
+
+export interface FieldStateProps extends TextInputProps {
+  validateOnStart?: boolean;
```
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/incubator/TextField/useFieldState.ts:9]: see my PR description. 
These props declared twice
In the main component (index.tsx) they are declared with doc comments

File: src/incubator/TextField/useFieldState.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+}: FieldStateProps) {
+  const [value, setValue] = useState(props.value);
+  const [isFocused, setIsFocused] = useState(false);
+  const [isValid, setIsValid] = useState(true);
+
+  useEffect(() => {
+    if (validateOnStart) {
+      validateField();
+    }
+  }, []);
```
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/incubator/TextField/useFieldState.ts:33]: I don't want this to run when these props change. 
useEffect with an empty dependencies array is similar to componentDidMount, which is what I wanted to achieve.

File: src/incubator/TextField/useFieldState.ts (hunk 1 above)
Comment (ethanshar) [reply src/incubator/TextField/useFieldState.ts:80]: no. 
This is the field state. 
It contains the field value, validity, focus state, etc...

The main component use the useFieldState and pass it on to various sub components, each using this data for different reasons.
For instance the ValidationMessage needs to know if the field is valid or not.. 
The Label needs to know when we're focus on the field for color calculation
and so on..

File: src/incubator/TextField/useFieldState.ts (hunk 2 above)
Comment (ethanshar) [reply src/incubator/TextField/useFieldState.ts:26]: Yes of course. 
isFocused affect the field UI and other functionality. 

For example:
The Field's label changes its color depending on the focus state
A validator can triggered when blurring the field and as a result show an validation error message.. 

It has to trigger a re-render.

File: src/incubator/TextField/useFieldState.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      let _isValid = true;
+      if (_.isFunction(validate)) {
+        _isValid = validate(valueToValidate);
+      } else if (_.isString(validate)) {
+        _isValid = _.invoke(validators, validate, valueToValidate);
+      }
+
+      setIsValid(_isValid);
+    },
+    [value]
```
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/incubator/TextField/useFieldState.ts]: done

File: src/incubator/TextField/useFieldState.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  const onChangeText = useCallback(
+    (text) => {
+      setValue(text);
+      _.invoke(props, 'onChangeText', text);
+
+      if (validateOnChange) {
+        validateField(text);
+      }
+    },
+    [props.onChangeText, validateOnChange]
```
[the remark this answers is not in the document]
Comment (ethanshar) [reply src/incubator/TextField/useFieldState.ts]: done

File: src/incubator/TextField/useFieldState.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/incubator/TextField/useFieldState.ts:9]: Yes, I noticed that, I think I missed the fact that `useFieldState` is not exported to the user.

**Review by M-i-k-e-l — APPROVED** [review]

**Pull request merged**